pkg/trace/agent/agent_test.go

```diff
@@ -758,126 +758,141 @@ func TestClientComputedStats(t *testing.T) {
+				{trace: generateProcessedTrace(sampler.PriorityAutoDrop, false), wantSampled: false},
+			},
 		},
 		"prio-sampled": {
-			hasPriority:     true,
-			prioritySampled: true,
-			wantSampled:     true,
+			agentConfig: agentConfig{},
+			testCases: []samplingTestCase{
+				{trace: generateProcessedTrace(sampler.PriorityAutoKeep, false), wantSampled: true},
+			},
 		},
 		"error-unsampled": {
-			hasErrors:     true,
-			errorsSampled: false,
-			wantSampled:   false,
+			agentConfig: agentConfig{errorsSampled: false},
+			testCases: []samplingTestCase{
+				{trace: generateProcessedTrace(sampler.PriorityNone, true), wantSampled: false},
+			},
 		},
 		"error-sampled": {
-			hasErrors:     true,
-			errorsSampled: true,
-			wantSampled:   true,
+			agentConfig: agentConfig{errorsSampled: true},
+			testCases: []samplingTestCase{
+				{trace: generateProcessedTrace(sampler.PriorityNone, true), wantSampled: true},
+			},
 		},
 		"error-sampled-prio-unsampled": {
-			hasErrors:       true,
-			hasPriority:     true,
-			errorsSampled:   true,
-			prioritySampled: false,
-			wantSampled:     true,
+			agentConfig: agentConfig{errorsSampled: true},
+			testCases: []samplingTestCase{
+				{trace: generateProcessedTrace(sampler.PriorityAutoDrop, true), wantSampled: true},
+			},
 		},
-		"error-unsampled-prio-sampled": {
-			hasErrors:       true,
-			hasPriority:     true,
-			errorsSampled:   false,
-			prioritySampled: true,
-			wantSampled:     true,
+		"error-sampled-prio-sampled": {
+			agentConfig: agentConfig{errorsSampled: false},
+			testCases: []samplingTestCase{
+				{trace: generateProcessedTrace(sampler.PriorityAutoKeep, true), wantSampled: true},
+			},
 		},
 		"error-prio-sampled": {
-			hasErrors:       true,
-			hasPriority:     true,
-			errorsSampled:   true,
-			prioritySampled: true,
-			wantSampled:     true,
+			agentConfig: agentConfig{errorsSampled: true},
+			testCases: []samplingTestCase{
+				{trace: generateProcessedTrace(sampler.PriorityAutoKeep, true), wantSampled: true},
+			},
 		},
 		"error-prio-unsampled": {
-			hasErrors:       true,
-			hasPriority:     true,
-			errorsSampled:   false,
-			prioritySampled: false,
-			wantSampled:     false,
+			agentConfig: agentConfig{errorsSampled: false},
+			testCases: []samplingTestCase{
+				{trace: generateProcessedTrace(sampler.PriorityAutoDrop, true), wantSampled: false},
+			},
 		},
-		"rare-sampler-catch": {
-			hasPriority: true,
-			wantSampled: true,
+		"rare-sampler-catch-unsampled": {
+			agentConfig: agentConfig{},
+			testCases: []samplingTestCase{
+				{trace: generateProcessedTrace(sampler.PriorityAutoDrop, false), wantSampled: true},
+			},
+		},
+		"rare-sampler-catch-sampled": {
+			agentConfig: agentConfig{},
+			testCases: []samplingTestCase{
+				{trace: generateProcessedTrace(sampler.PriorityAutoKeep, false), wantSampled: true},
+				{trace: generateProcessedTrace(sampler.PriorityAutoDrop, false), wantSampled: false},
```

Comment:
This whole refactoring is due to this test, the bug fixed by this PR was only reproducible by sending 2 chunks to the same agent object.